### ISSUE TITLE
Random UCX Errors in ATDM CUDA 10.1 builds on 'white' and 'ride'

### ISSUE THREAD

**Issue by bartlettroscoe** [trilinos/Trilinos#5713]
@trilinos/anasazi, @trilinos/belos, @trilinos/muelu, @trilinos/nox, @trilinos/teko, @trilinos/tempus, @srajama1 (Trilinos Linear Solvers Product Lead), @rppawlo (Trilinos Nonlinear Solvers Product Lead), @fryeguy52 

## Next Action Item

Still seeing random failures for a few tests showing this UCX error as of 8/14/2019.  Next: Wait for fix to UCX and extend `cdash_analyze_and_report.py` tool to filter tests that shows this error message and associate them with this issue ...

## Description

This issue is an aggregate of the issues #5033, #5035, #5221, #5222, and #5224

As shown in [this query](https://testing-dev.sandia.gov/cdash/queryTests.php?project=Trilinos&date=2019-08-13&filtercount=6&showfilters=1&filtercombine=and&field1=buildname&compare1=61&value1=Trilinos-atdm-white-ride-cuda-10.1-gnu-7.2.0-release-debug&field2=site&compare2=61&value2=ride&field3=status&compare3=62&value3=passed&field4=details&compare4=64&value4=Timeout&field5=buildstarttime&compare5=83&value5=2019-06-15&field6=buildstarttime&compare6=84&value6=2019-08-15) in the 60 days from 6/15/2019 through 8/14/2019:

| Test Name | Number of failures |
| :--       |                --: |
| Anasazi_BlockKrylovSchurThyra_test_MPI_4 | 4 |
| Anasazi_Epetra_BKS_nh_test_0_MPI_4 | 1 |
| Anasazi_Epetra_BKS_solvertest_MPI_4 | 1 |
| Anasazi_Epetra_BKS_test_0_MPI_4 | 2 |
| Anasazi_Epetra_GeneralizedDavidson_nh_test_MPI_4 | 6 |
| Anasazi_Epetra_OrthoManagerMatTester_0_MPI_4 | 4 |
| Anasazi_Epetra_OrthoManagerTester_1_MPI_4 | 1 |
| Anasazi_GeneralizedDavidsonThyra_test_MPI_4 | 1 |
| Belos_bl_fgmres_hb_MPI_4 | 2 |
| Belos_bl_pgmres_hb_0_MPI_4 | 9 |
| Belos_bl_pgmres_hb_1_MPI_4 | 1 |
| Belos_bl_pgmres_hb_2_MPI_4 | 10 |
| Belos_BlockGmresPoly_Epetra_File_Ex_2_MPI_4 | 1 |
| Belos_gcrodr_hb_MPI_4 | 1 |
| Belos_pseudo_pgmres_hb_MPI_4 | 2 |
| Belos_resolve_gmres_hb_1_MPI_4 | 1 |
| MueLu_Maxwell3D-Epetra_MPI_4 | 23 |
| NOX_1DfemStratimikos_MPI_4 | 1 |
| NOX_1DfemStratimikosInexact_MPI_4 | 1 |
| NOX_LOCA_Tcubed_TurningPoint_Example_MPI_1 | 1 |
| NOX_LOCA_TcubedTP_MPI_2 | 7 |
| NOX_LOCA_TcubedTP_stratimikos_MPI_2 | 10 |
| Teko_testdriver_MPI_4 | 48 |
| Tempus_BackwardEuler_MPI_1 | 4 |
| Tempus_BDF2_MPI_1 | 4 |

in the build:

* Trilinos-atdm-white-ride-cuda-10.1-gnu-7.2.0-release-debug

on 'ride'.  (There is also the same build running on 'white' as well that also shows some of these failures.)

Looking at many of these failures they abort with errors like:

```
Caught signal 11 (Segmentation fault: address not mapped to object at address <memory-address>)
```

(NOTE: Since there are 146 failing tests we did not look at all of the test output to verify they all filed with this error.  We have asked Kitware to add an the ability to filter tests based on regexes of the test output which would allow us to determine for sure which tests are failing with this error (see [this Kitware GitLab Issue](https://gitlab.kitware.com/snl/project-1/issues/103)).)

It was previously determined from triaging the issues #5033, #5035, #5221, #5222, and #5224 that these random failures are likely due to a defect in a low library called UCX with CUDA 10.1 which is used by OpenMPI to manage CUDA memory.

Note that while these errors are somewhat random, only a small subset of the Trilinos tests have shown any failures and some tests show much more frequent failures like MueLu_Maxwell3D-Epetra_MPI_4 (failed 23 out of 60 days) and Teko_testdriver_MPI_4 (failed 48 out of 60 days).  Therefore, there must be some usage pattern used in some of these tests that triggers this error more.

## Current status on CDash:

The failing tests on this build for the last 7 days can be seen in [this query](https://testing-dev.sandia.gov/cdash/queryTests.php?project=Trilinos&date=2019-08-13&filtercount=5&showfilters=1&filtercombine=and&field1=buildname&compare1=61&value1=Trilinos-atdm-white-ride-cuda-10.1-gnu-7.2.0-release-debug&field2=site&compare2=61&value2=ride&field3=status&compare3=62&value3=passed&field4=details&compare4=64&value4=Timeout&field5=buildstarttime&compare5=83&value5=7%20days%20ago).

## Steps to Reproduce

One should be able to reproduce these failures on 'ride' or 'white' as described in:

* https://github.com/trilinos/Trilinos/blob/develop/cmake/std/atdm/README.md

More specifically, the commands given for 'ride' or 'white' are provided at:

* https://github.com/trilinos/Trilinos/blob/develop/cmake/std/atdm/README.md#ridewhite

The exact commands to reproduce this issue should be:

```
$ cd <some_build_dir>/
$ source $TRILINOS_DIR/cmake/std/atdm/load-env.sh Trilinos-atdm-white-ride-cuda-10.1-gnu-7.2.0-release-debug
$ cmake \
 -GNinja \
 -DTrilinos_CONFIGURE_OPTIONS_FILE:STRING=cmake/std/atdm/ATDMDevEnv.cmake \
 -DTrilinos_ENABLE_TESTS=ON \
 -DTrilinos_ENABLE_Anasazi=ON \
 -DTrilinos_ENABLE_Belos=ON \
 -DTrilinos_ENABLE_MueLu=ON \
 -DTrilinos_ENABLE_NOX=ON \
 -DTrilinos_ENABLE_Teko=ON \
 -DTrilinos_ENABLE_Tempus=ON \
 $TRILINOS_DIR

$ make NP=16

$ bsub -x -Is -q rhel7F -n 16 ctest -j16
```

**Comment by bartlettroscoe** [comment]
The best we know is that this is caused by an issue in the UCX library used in OpenMPI.  Therefore, I have put the `ATDM Env Issue` label on this Issue so this can be filtered out of GitHub issue queries.

I will now close the issues #5033, #5035, #5221, #5222, and #5224 as duplicates of this issue.

**Comment by bartlettroscoe** [comment]
FYI: As noted in [TRIL-245](https://sems-atlassian-son.sandia.gov/jira/browse/TRIL-245), a new CUDA 10.1 env will be getting set up soon on 'waterman' that might resolve these UCX errors.

**Comment by prwolfe** [comment]
We just saw a failure of MueLu_Maxwell3D-Tpetra-Stratimikos_MPI_4 in the master-merge branch using gcc 4.8.4. This tends to point toward non-cuda issues for this test.

https://testing.sandia.gov/cdash/testDetails.php?test=87994601&build=5569462

**Comment by bartlettroscoe** [comment]
@prwolfe said:

> This tends to point toward non-cuda issues for this test.
>
> https://testing.sandia.gov/cdash/testDetails.php?test=87994601&build=5569462

If you look at that test output, it does not show the "Segmentation fault: ... object at address <memory-address>" errors that characterize these OpenMPI/UCX errors.  Instead, that test showed errors like:

```
p=2: *** Caught standard std::exception of type 'std::runtime_error' :

 /scratch/trilinos/workspace/trilinos-folder/Trilinos_pullrequest_gcc_4.8.4/Trilinos/packages/amesos2/src/Amesos2_KLU2_def.hpp:277:
 
 Throw number = 1
 
 Throw test that evaluated to true: info > 0
 
 KLU2 numeric factorization failed
```

Therefore, I think a new issue is required for this new type of failure for the test  MueLu_Maxwell3D-Tpetra-Stratimikos_MPI_4.

**Comment by cgcgcg** [comment]
I believe that Klu2 was recently changed to give better output when trying to solve a singular system. I'll have to investigate why this is happening in this case.

**Comment by bartlettroscoe** [comment]
FYI: The failing anasazi tests associated with this issue have been disabled as per:

* https://sems-atlassian-srn.sandia.gov/browse/ATDV-263

**Comment by bartlettroscoe** [comment]
FYI: As shown in [this query](https://testing-dev.sandia.gov/cdash/queryTests.php?project=Trilinos&begin=2020-01-14&end=2020-02-12&filtercount=3&showfilters=1&filtercombine=and&field1=buildname&compare1=61&value1=Trilinos-atdm-white-ride-cuda-10.1-gnu-7.2.0-release-debug&field2=testname&compare2=61&value2=Belos_pseudo_pgmres_hb_MPI_4&field3=site&compare3=61&value3=ride), also seeing these failures in the test:

* `Belos_pseudo_pgmres_hb_MPI_4`

in the build:

* `Trilinos-atdm-white-ride-cuda-10.1-gnu-7.2.0-release-debug`

showing errors like:

```
[ride27:85735:0:85735] Caught signal 11 (Segmentation fault: address not mapped to object at address 0x1002c3f0050)
```

**Comment by bartlettroscoe** [comment]
Now that we have a real ATS-2 env (almost) correctly running the native Trilinos test suite on 'vortex', why are we continuing to bother with this experimental CUDA 10.1 + OpenMPI 4.0.1 build that exhibits random UCX failures?  What is the point?

**Comment by bartlettroscoe** [comment]
> Now that we have a real ATS-2 env (almost) correctly running the native Trilinos test suite on 'vortex', why are we continuing to bother with this experimental CUDA 10.1 + OpenMPI 4.0.1 build that exhibits random UCX failures? What is the point?

I asked this question at the ATDM PI meeting today and Mike Glass's opinion is that we should not bother with this experimental environment at this point.  Therefore, I am going to shut down this ATDM Trilinos build and after that I will close this issue.

In the future, I think we need to be more careful about what experimental builds we put resources into trying to clean up.  Once we discovered that there were system issues with this version of CUA, OpenMPI, and UCX, we likely should have just shut this down.

**Comment by bartlettroscoe** [comment]
I removed the nightly build:

* Trilinos-atdm-white-ride-cuda-10.1-gnu-7.2.0-release-debug

and all tracking of failures allocated with that build.  See:

* https://sems-atlassian-srn.sandia.gov/browse/ATDV-308

I am closing this as "wontfix".  I am not adding the label "Disabled Tests" since there is nothing that Trilinos developers could have done to fix this. This was a bug in the underlying system software.

Again, in the future, I think we need to be more careful about what experimental builds we put resources into trying to clean up. Once we discovered that there were system issues with this version of CUA, OpenMPI, and UCX, we likely should have just shut this down.
